tidy3d/plugins/design/design.py

```diff
@@ -86,38 +116,281 @@ def get_fn_source(function: Callable) -> str:
+        )
+        fn_args, fn_values, aux_values, opt_output = self.method._run(
+            run_fn=handler.fn_combined, parameters=self.parameters, console=console
+        )
+        return fn_args, fn_values, aux_values, opt_output, handler.sim_names, handler.sim_paths
+
+    """ Helpers """
+
+    def _get_evaluate_fn_single(self, fn: Callable) -> list[Any]:
+        """Get function that sequentially evaluates single `fn` for a list of arguments."""
+
+        def evaluate(args_list: list) -> list[Any]:
+            """Evaluate a list of arguments passed to ``fn``."""
+            return [fn(**args) for args in args_list]
+
+        return evaluate
+
+    def _get_evaluate_fn_pre_post(
+        self, fn_pre: Callable, fn_post: Callable, fn_mid: Callable, verbose: bool
+    ):
+        """Get function that tries to use batch processing on a set of arguments."""
+
+        class Pre_Post_Handler:
+            def __init__(self, verbose):
+                self.sim_counter = 0
+                self.sim_names = []
+                self.sim_paths = []
+                self.verbose = verbose
+
+            def fn_combined(self, args_list: list[dict[str, Any]]) -> list[Any]:
+                """Compute fn_pre and fn_post functions and capture other outputs."""
+                sim_dict = {str(idx): fn_pre(**arg_list) for idx, arg_list in enumerate(args_list)}
+
+                if not all(
+                    isinstance(val, (int, float, Simulation, Batch, list, dict))
+                    for val in sim_dict.values()
+                ):
+                    raise ValueError(
+                        "Unrecognized output of fn_pre. Please change the return of fn_pre."
+                    )
+
+                data, task_names, task_paths, sim_counter = fn_mid(
+                    sim_dict, self.sim_counter, self.verbose
+                )
+                self.sim_names.extend(task_names)
+                self.sim_paths.extend(task_paths)
+                self.sim_counter = sim_counter
+                post_out = [fn_post(val) for val in data.values()]
+                return post_out
+
+        handler = Pre_Post_Handler(verbose)
+
+        return handler
+
+    def _fn_mid(
+        self, pre_out: dict[int, Any], sim_counter: int, verbose: bool
+    ) -> Union[dict[int, Any], BatchData]:
+        """A function of the output of ``fn_pre`` that gives the input to ``fn_post``."""
+
+        # Keep copy of original to use if no tidy3d simulation required
+        original_pre_out = pre_out.copy()
+
+        # Convert list input to dict of dict before passing through checks
+        was_list = False
+        if all(isinstance(sim, list) for sim in pre_out.values()):
+            pre_out = {
+                str(idx): {str(sub_idx): sub_list for sub_idx, sub_list in enumerate(sim_list)}
+                for idx, sim_list in enumerate(pre_out.values())
+            }
+            was_list = True
+
+        def _find_and_map(
```

Comment:
this is quite complex :D . but I think we discussed it enough. if it's understandable enough to you it's fine. And we can always work on it in the future

Reply:
It certainly is. Do you think it would help moving all the inline func definitions to the top of the file, or to a `design_utils.py`?

Reply:
hm, I think it might make sense to leave them here actually. since they're easier to reference when reading through. 

some of this complexity is probably unavoidable..